crates/trie/parallel/src/proof_task.rs

```diff
@@ -0,0 +1,333 @@
+    total_transactions: usize,
+    /// Consistent view provider used for creating transactions on-demand
+    view: ConsistentDbView<Factory>,
+    /// Proof task context shared across all proof tasks
+    task_ctx: ProofTaskCtx,
+    /// Storage proof input pending execution
+    pending_proofs: VecDeque<(StorageProofInput, Sender<StorageProofResult>)>,
+    /// The underlying handle from which to spawn proof tasks
+    executor: Handle,
+    /// The proof task transactions, containing owned cursor factories that are reused for proof
+    /// calculation.
+    proof_task_txs: Vec<ProofTaskTx<FactoryTx<Factory>>>,
+    /// A receiver for new proof tasks.
+    proof_task_rx: Receiver<ProofTaskMessage<FactoryTx<Factory>>>,
+    /// A sender for sending back transactions.
+    tx_sender: Sender<ProofTaskMessage<FactoryTx<Factory>>>,
+}
+
+impl<Factory: DatabaseProviderFactory> ProofTaskManager<Factory> {
+    /// Creates a new [`ProofTaskManager`] with the given max concurrency, creating that number of
+    /// cursor factories.
+    ///
+    /// Returns an error if the consistent view provider fails to create a read-only transaction.
+    pub fn new(
+        executor: Handle,
+        view: ConsistentDbView<Factory>,
+        task_ctx: ProofTaskCtx,
+        max_concurrency: usize,
+    ) -> Self {
+        let (tx_sender, proof_task_rx) = channel();
+        Self {
+            max_concurrency,
+            total_transactions: 0,
+            view,
+            task_ctx,
+            pending_proofs: VecDeque::new(),
+            executor,
+            proof_task_txs: Vec::new(),
+            proof_task_rx,
+            tx_sender,
+        }
+    }
+
+    /// Returns a handle for sending new proof tasks to the [`ProofTaskManager`].
+    pub fn handle(&self) -> ProofTaskManagerHandle<FactoryTx<Factory>> {
+        ProofTaskManagerHandle::new(self.tx_sender.clone())
+    }
+}
+
+impl<Factory> ProofTaskManager<Factory>
+where
+    Factory: DatabaseProviderFactory<Provider: BlockReader> + StateCommitmentProvider + 'static,
+{
+    /// Inserts the storage proof input and sender into the pending proof queue.
+    pub fn queue_proof_task(
+        &mut self,
+        input: StorageProofInput,
+        sender: Sender<StorageProofResult>,
+    ) {
+        self.pending_proofs.push_back((input, sender));
+    }
+
+    /// Gets either the next available transaction, or creates a new one if all are in use and the
+    /// total number of transactions created is less than the max concurrency.
+    pub fn get_or_create_tx(&mut self) -> ProviderResult<Option<ProofTaskTx<FactoryTx<Factory>>>> {
+        if let Some(proof_task_tx) = self.proof_task_txs.pop() {
+            return Ok(Some(proof_task_tx));
+        }
+
+        // if we can create a new tx within our concurrency limits, create one on-demand
+        if self.total_transactions < self.max_concurrency {
+            let provider_ro = self.view.provider_ro()?;
```

Comment:
this we could potentially also do in the task but since we're limiting the number of txs we only have to this the first 256 times, so is likely fine